### ISSUE TITLE
Reproduce Profilememory Interactive Testing Results of Pretrained Model

### ISSUE THREAD

**Issue by christqoh** [facebookresearch/ParlAI#967]
(Related to #823)
When training the Profilememory from scratch and interactively testing it I get stuff like
`[Seq2Seq]: i ' ve a lot of free time . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . . . __UNK__ . . . __UNK__ . . . . .`

where validation during training looks like this:
`prediction: i ' m a student __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__ __START__`
any suggestions where that might originate in?

**Comment by alexholdenmiller** [comment]
we did change the default tokenizer used for the train scripts, it's possible you ran the script before and built the dictionary using the old tokenizer and then are now training using the new one without rebuilding the dictionary, which could create problems with mapping the tokens together.

(Loading the pretrained models would override the tokenizer option so should be no issues there.)

Just to be safe, I just put up #968 to set the default tokenizer in the train scripts to the old tokenizer.

**Comment by christqoh** [comment]
Do I not rebuild the dictionary every time I train from scratch?

When interactively testing your pretrained model I run 
`python projects/convai2/baselines/profilememory/interactive.py`. 
When I want to test mine I add the according parser args for the model and dictionary, so 
`python projects/convai2/baselines/profilememory/interactive.py --model-file [location of model file] --dict-file [location of dict file]`. 
 I noticed that the dict file of the pretrained model and mine diverge, where for instance the pretrained does not include `\n` while one trained by me does. 

So consequently I will have to include a tokenizer parser argument in both training and interacting call to make sure things are compatible? Could you quickly elaborate on the tokenizer and its arguments, please?

**Comment by christqoh** [comment]
just tried by training from scratch with `dict_tokenizer='split'` set in the training script. Did not work for the interactive testing with call as shown above. Still have the UNKs in there. Quite a few. Stopped predicting fullstops between though.

**Comment by alexholdenmiller** [comment]
The dictionary is not rebuilt if the file exists already, as on longer datasets it can take quite a while to set up.

The presence/lack of the `\n` show that you are using the new default tokenizer, `-tok re`. If the dictionary was previously built with `-tok split` (the previous default) and was not rebuilt using the new one, you'll get mismatches and a lot of UNKs, as you did.

If you delete all the files created by the training script (model & dictionary files) and start over it should work!

**Comment by christqoh** [comment]
As I understood, during training, files are saved to `/tmp/`. I train through running a shell script. First line in there cleans up the /tmp/ folder as I had issues with historic files before.

I noticed that even though I included `dict_tokenizer='split'` in my training script, `\n`s are missing and I get the UNKs when interactively testing. 

As `model_file='/tmp/profilemem_convai2'` is explicitly stated in the training script, I suppose I have to do the same for the dict file to not accidentally load the one of the pretrained model?!

**Comment by alexholdenmiller** [comment]
Yes when split is used for the tokenizer, newlines will not be in the dictionary. As far as UNK's when interactively testing, that is possible if you type in out-of-vocabulary words during interactive mode, but are not expected during training validation.

No, the pretrained model doesn't put its dictionary in /tmp/. I would just say to make sure you delete the model file and dictionary file before retraining from scratch right now.

**Comment by christqoh** [comment]
The pretrained stuff is located somewhere else, that's true. I `rm -rf /tmp/` so I don't expect temporary files to be the issue... Maybe, do I need to include dict_tokenizer specifications in the interactive mode?

**Comment by alexholdenmiller** [comment]
You could add that to make sure!

**Comment by christqoh** [comment]
so I trained a profilememory from scratch again, cleared the `/tmp/` before, explicitly pointed `dict_file='/tmp/profilemem_convai2.dict'` to `/tmp/` to avoid loading from existing dictionary, trained with `dict_tokenizer='split'` set. Training validation again gave lots of `__START__`s as mentioned in first post. 

Tried interacting with and without `--dict_tokenizer split` set ... results are simply equivalent. 

`Enter Your Message: moin moin`
`[Seq2Seq]: i am not sure what you mean . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__ . __UNK__`

Do you have any further ideas how I get to a proper interaction?

**Comment by alexholdenmiller** [comment]
Since you trained the model with no unknown tokens during training, I'm not surprising that is said nonsense when you said "`__UNK__ __UNK__`" to it (as it would have interpreted "moin moin", since they aren't in the dictionary). Given two random input vectors, "i am not sure what you mean" is actually the perfect response!

Unfortunately though I'm not exactly sure what's going wrong during your training. One option is to switch over to the seq2seq agent, which does attention over the persona but does not separate it out from the context as the persona_seq2seq model does. The regular seq2seq model also just never outputs start tokens--it strips them out when converting vectors to text--which can help clean up extraneous ones at the end of the output.

**Comment by christqoh** [comment]
sorry for that, not an ideal example from my side. Same model, different input:
`Enter Your Message: hey there, how are you?`
`[Seq2Seq]: i'm good . just got done walking my dog . __UNK__ you ? __UNK__ you ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__ ? __UNK__`
so it understands what I say, I never claimed otherwise, it just somehow didn't learn the length of the response sequence. 
I will look into it again and come back to you.

**Comment by alexholdenmiller** [comment]
ah yeah that's definitely bad behavior. strange for it to produce UNK instead of END at the end of its sequence, but I don't have any good explanation for that.

**Comment by christqoh** [comment]
So I just did a clean pull, cleared up my `/tmp/` and trained running `python projects/convai2/baselines/profilememory/train.py &> log.txt` only (compare here: [log.txt](https://github.com/facebookresearch/ParlAI/files/2185423/log.txt)). Doesn't seem to be any better. Are you able to reproduce this behaviour?

When trying to run `python projects/convai2/baselines/profilememory/interactive.py --model-file /tmp/profilememconvai2 --dict-file /tmp/profilememconvai2.dict --dict_tokenizer spli
t` in that clean version  I noticed that it starts downloading the pretrained model. Any ideas why that happens?

**Comment by alexholdenmiller** [comment]
As far as downloading the pretrained model, yeah it always just downloads it whether you're using your own or not. We could add a check for the model filename there, but haven't, as it was originally set up for you to interact with the pretrained model.

Okay so we can see the model is actually learning to do this bad behavior; during the training, we see this log:
```
prediction: i ! ! . i you have any ? __UNK__ a dogs . a . i __UNK__ __UNK__ a . . . . . . . . . . . 
label: hi there greg . do you have pets ? i've 2 cats named milo and fio . __UNK__ __START__ __START__ __START__ __START__ __START__ __START__
```

This means that the model is accidentally using `__UNK__` where `__END__` should be and `__START__` where `__NULL__` should be. This is reproducing for me, so I'll take a look at fixing it. Thanks for catching!

**Comment by alexholdenmiller** [comment]
Okay I found it! There was an issue where the NULL index (supposed to be `0`) was being set by having the dictionary *parse* the NULL token (`__NULL__`), but since dict-lower was being set to true it was seeing `__null__` and marking it as an unknown token. Thus, it was not putting the proper padding in.

Then, as far as the UNK being sown at the end of the sentence, it is the same thing: it's parsing "milo and fio . __end__" and the lowercase is making it print __UNK__. however, this is just a logging "bug"--it should still be training correctly at that point.

I'm going to make some adjustments to the model and open a PR that should help with these issues.

**Comment by alexholdenmiller** [comment]
Merged fix, closing for now but feel free to reopen!
